haskell/private/cc_wrapper.py.tpl

```diff
@@ -0,0 +1,812 @@
+compiler.
+
+- Shortens library search paths to stay below maximum path length on Windows.
+
+    GHC generates library search paths that contain redundant up-level
+    references (..). This can exceed the maximum path length on Windows, which
+    will cause linking failures. This wrapper shortens library search paths to
+    avoid that issue.
+
+- Shortens rpaths and load commands on macOS.
+
+    The rpaths and load commands generated by GHC and Bazel can quickly exceed
+    the MACH-O header size limit on macOS. This wrapper shortens and combines
+    rpaths and load commands to avoid exceeding that limit.
+
+- Finds .so files if only .dylib are searched on macOS.
+
+    Bazel's cc_library will generate .so files for dynamic libraries even on
+    macOS. GHC strictly expects .dylib files on macOS. This wrapper hooks into
+    gcc's --print-file-name feature to work around this mismatch in file
+    extension.
+
+"""
+
+from bazel_tools.tools.python.runfiles import runfiles as bazel_runfiles
+from contextlib import contextmanager
+import glob
+import itertools
+import os
+import platform
+import shlex
+import subprocess
+import sys
+import tempfile
+
+CC = "{:cc:}"
+INSTALL_NAME_TOOL = "/usr/bin/install_name_tool"
+OTOOL = "/usr/bin/otool"
+
+
+def main():
+    parsed = Args(load_response_files(sys.argv[1:]))
+
+    if parsed.linking:
+        link(parsed.output, parsed.libraries, parsed.rpaths, parsed.args)
+    elif parsed.printing_file_name:
+        print_file_name(parsed.print_file_name, parsed.args)
+    else:
+        run_cc(parsed.args, exit_on_error=True)
+
+
+# --------------------------------------------------------------------
+# Parse arguments
+
+
+class Args:
+    """Parsed command-line arguments.
+
+    Attrs:
+      args: The collected and transformed arguments.
+
+      linking: The action is linking.
+      printing_file_name: The action is print-file-name.
+
+      output: The output binary or library when linking.
+      library_paths: The library search paths when linking.
+      libraries: The required libraries when linking.
+      rpaths: The provided rpaths when linking.
+
+      print_file_name: The queried file name on print-file-name.
+
+    """
+    LINK = "link"
+    COMPILE = "compile"
+    PRINT_FILE_NAME = "print-file-name"
+
+    def __init__(self, args):
+        """Parse the given arguments into an Args object.
+
+        - Shortens library search paths.
+        - Detects the requested action.
+        - Keeps rpath arguments for further processing when linking.
+        - Keeps print-file-name arguments for further processing.
+
+        Args:
+          args: Iterable over command-line arguments.
+
+        """
+        self.action = Args.LINK
+        self.print_file_name = None
+        self.libraries = []
+        self.library_paths = []
+        self.rpaths = []
+        self.output = None
+        self._prev_ld_arg = None
```

Comment:
These members could do with some docstrings

Reply:
They are defined in the `Attrs` section of the class docstring [here](https://github.com/tweag/rules_haskell/pull/1039/files/71c2a6d5286a0e0802c78be3a1a8cf6e6ae8061d..a10b82374bc7b6f86844835b1656e06b449a0b0a#diff-19cb794875a66149f66f34681dd62136R66).